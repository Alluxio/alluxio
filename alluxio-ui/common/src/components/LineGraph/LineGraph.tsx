/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

import {Theme} from '@nivo/core';
import {LineSerieData, ResponsiveLine} from '@nivo/line';
import React from 'react';

import './LineGraph.css';

export interface ILineGraphProps {
  data: LineSerieData[];
  xAxisLabel: string;
  xAxisUnits: string;
  yAxisLabel: string;
  yAxisUnits: string;
}

export class LineGraph extends React.PureComponent<ILineGraphProps> {
  public render(): JSX.Element {
    const {data, xAxisLabel, yAxisLabel} = this.props;
    const nivoTheme: Theme = {
      background: 'transparent',
      axis: {
        domain: {line: {stroke: 'transparent', strokeWidth: 1}},
        ticks: {line: {stroke: '#fff', strokeWidth: 1}, text: {fill: '#bbb', fontSize: 12}},
        legend: {text: {fill: '#bbb', fontSize: 12}}
      },
      grid: {line: {stroke: '#ddd', strokeWidth: 1}},
      legends: {text: {fill: '#bbb', fontSize: 12}},
      labels: {text: {fill: '#bbb', fontSize: 12}},
      markers: {lineColor: '#fff', lineStrokeWidth: 1, textColor: '#999', fontSize: '12'},
      dots: {text: {fill: '#bbb', fontSize: 12}},
      tooltip: {
        container: {
          background: '#222',
          color: 'inherit',
          fontSize: 'inherit',
          borderRadius: '2px',
          boxShadow: '0 1px 2px rgba(255, 255, 255, 0.25)',
          padding: '5px 9px'
        }, basic: {whiteSpace: 'pre', display: 'flex', alignItems: 'center'}, table: {}, tableCell: {padding: '3px 5px'}
      }
    };

    return (
      <div className="lineGraph col-12 col-md-6 col-xl-4">
        <h5>{data[0].id}</h5>
        <ResponsiveLine
          theme={nivoTheme}
          data={data}
          margin={{top: 25, right: 50, bottom: 200, left: 75}}
          xScale={{type: 'time', format: '%Q', precision: 'minute'}}
          yScale={{type: 'linear', stacked: false, min: 0, max: 100}}
          axisTop={null}
          axisRight={null}
          axisBottom={{
            tickSize: 5,
            tickPadding: 5,
            tickRotation: -90,
            legend: xAxisLabel,
            legendOffset: 140,
            legendPosition: 'middle',
            format: '%x %H:%M:%S'
          }}
          axisLeft={{
            tickSize: 5,
            tickPadding: 5,
            tickRotation: 0,
            legend: yAxisLabel,
            legendOffset: -40,
            legendPosition: 'middle'
          }}
          dotSize={8}
          dotColor='inherit:darker(0.3)'
          dotBorderWidth={1}
          dotBorderColor="#fff"
          enableDotLabel={true}
          animate={true}
          motionStiffness={90}
          motionDamping={15}
          curve='natural'
          colors='set3'
          enableArea={true}
          areaOpacity={0.5}
          areaBaselineValue={0}
          isInteractive={true}
          enableStackTooltip={true}
        />
      </div>
    );
  }
}
