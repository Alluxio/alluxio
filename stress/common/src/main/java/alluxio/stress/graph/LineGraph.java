/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

package alluxio.stress.graph;

import com.google.common.collect.ImmutableMap;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * A line graph representation.
 */
public class LineGraph extends Graph {
  private static final String X_FIELD = "x";
  private static final String Y_FIELD = "y";
  private static final String SERIES_FIELD = "series";

  private static final int POINT_SIZE = 81;
  private static final Object CONDITIONAL_OPACITY = ImmutableMap.of("opacity", ImmutableMap
      .of("condition", ImmutableMap.of("selection", "legendSelect", "value", 1), "value", 0.05));
  private static final Object LEGEND_CONFIG = ImmutableMap.of("columns", 3, "symbolLimit", 500);

  /**
   * This represents the data for a line graph.
   */
  public static final class Data {
    private final List<Map<Object, Object>> mData;

    /**
     * Creates an instance.
     */
    public Data() {
      mData = new ArrayList<>();
    }

    /**
     * @param x the x value
     * @param y the y value
     */
    public void addData(Object x, Object y) {
      mData.add(ImmutableMap.of(LineGraph.X_FIELD, x, LineGraph.Y_FIELD, y));
    }
  }

  /**
   * Creates an instance.
   *
   * @param title the title
   * @param subTitle the lines in the sub title
   * @param xTitle the title of the x-axis
   * @param yTitle the title of the y-axis
   */
  public LineGraph(String title, List<String> subTitle, String xTitle, String yTitle) {
    mTitle.add(title);
    mSubTitle.addAll(subTitle);

    mGraph.put("$schema", "https://vega.github.io/schema/vega-lite/v4.json");
    mGraph.put("width", 700);
    mGraph.put("height", 400);
    mGraph.put("title", ImmutableMap.of("text", mTitle, "subtitle", mSubTitle));

    // TODO(gpang): is this the best we can do to build out this map/json?
    Map<String, Object> encoding = new HashMap<>();
    // TODO(gpang): customize the type
    encoding.put("x", ImmutableMap.of("field", X_FIELD, "type", "ordinal", "title", xTitle));
    encoding.put("y",
        ImmutableMap.of("field", Y_FIELD, "type", "quantitative", "title", yTitle));
    encoding.put("tooltip", Arrays
        .asList(ImmutableMap.of("field", SERIES_FIELD, "type", "nominal", "title", "Series"),
            ImmutableMap.of("field", X_FIELD, "type", "ordinal", "title", xTitle),
            ImmutableMap.of("field", Y_FIELD, "type", "quantitative", "title", yTitle)));
    encoding.put("color",
        ImmutableMap.of("field", SERIES_FIELD, "type", "nominal", "legend", LEGEND_CONFIG));
    encoding.put("strokeWidth", ImmutableMap.of("value", 3));
    mGraph.put("encoding", encoding);

    List<Object> layer = new ArrayList<>();
    // This layer is for the transparent points and the selection
    layer.add(ImmutableMap
        .of("mark", ImmutableMap.of("type", "point", "size", POINT_SIZE, "stroke", "transparent"),
            "selection", ImmutableMap.of("legendSelect", ImmutableMap
                .of("type", "multi", "bind", "legend", "fields", new String[] {SERIES_FIELD}))));
    // This layer is for the line (without points)
    layer.add(ImmutableMap.of("mark", "line", "encoding", CONDITIONAL_OPACITY));
    // This layer is for the points
    layer.add(ImmutableMap
        .of("mark",
            ImmutableMap.of("type", "point", "size", POINT_SIZE, "filled", false, "fill", "white"),
            "encoding", CONDITIONAL_OPACITY));
    mGraph.put("layer", layer);

    mGraph.put("config", ImmutableMap.of("legend",
        ImmutableMap.of("orient", "bottom", "direction", "vertical", "labelLimit", 700)));
  }

  /**
   * @param series the series name
   * @param data the series data to add
   */
  public void addDataSeries(String series, LineGraph.Data data) {
    List<Map<Object, Object>> newSeries =
        data.mData.stream().map(m -> {
          Map<Object, Object> newMap = new HashMap<>(m);
          newMap.put(LineGraph.SERIES_FIELD, series);
          return newMap;
        }).collect(Collectors.toList());
    mData.addAll(newSeries);
  }
}
