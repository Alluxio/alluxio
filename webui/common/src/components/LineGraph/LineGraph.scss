.lineGraph {
  height: 25em;
}
