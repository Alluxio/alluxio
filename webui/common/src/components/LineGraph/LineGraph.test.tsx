/*
 * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
 * (the "License"). You may not use this work except in compliance with the License, which is
 * available at www.apache.org/licenses/LICENSE-2.0
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied, as more fully set forth in the License.
 *
 * See the NOTICE file distributed with this work for information regarding copyright ownership.
 */

import { configure, shallow, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

import { ILineGraphProps, LineGraph } from './LineGraph';

configure({ adapter: new Adapter() });

describe('Paginator', () => {
  let props: ILineGraphProps;

  beforeAll(() => {
    props = {
      data: [{ id: '', data: [] }],
      yAxisLabel: '',
      xAxisLabel: '',
      xAxisUnits: '',
      yAxisUnits: '',
    };
  });

  describe('Shallow component', () => {
    let shallowWrapper: ShallowWrapper;

    beforeEach(() => {
      shallowWrapper = shallow(<LineGraph {...props} />);
    });

    it('Renders without crashing', () => {
      expect(shallowWrapper.length).toEqual(1);
    });

    it('Matches percent y-axis snapshot', () => {
      shallowWrapper.setProps({
        data: [
          {
            id: '% Test',
            data: [{ x: 1614803627951, y: 0 }],
          },
        ],
        yAxisLabel: 'Percent (%)',
        xAxisLabel: 'Time Stamp',
      });
      expect(shallowWrapper).toMatchSnapshot();
    });

    it('Matches throughput y-axis snapshot', () => {
      shallowWrapper.setProps({
        data: [
          {
            id: 'Test',
            data: [{ x: 1614803627951, y: 0 }],
          },
        ],
        yAxisLabel: 'Raw Throughput (Bytes/Minute)',
        xAxisLabel: 'Time Stamp',
      });
      expect(shallowWrapper).toMatchSnapshot();
    });
  });
});
