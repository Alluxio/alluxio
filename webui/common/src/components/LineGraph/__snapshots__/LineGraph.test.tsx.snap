// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator Shallow component Matches percent y-axis snapshot 1`] = `
<div
  className="lineGraph col-12 col-md-6 col-xl-4"
>
  <h5>
    % Test
  </h5>
  <ResponsiveLine
    animate={true}
    areaBaselineValue={0}
    areaOpacity={0.5}
    axisBottom={
      Object {
        "format": "%x %H:%M:%S",
        "legend": "Time Stamp",
        "legendOffset": 140,
        "legendPosition": "middle",
        "tickPadding": 5,
        "tickRotation": -90,
        "tickSize": 5,
      }
    }
    axisLeft={
      Object {
        "legend": "Percent (%)",
        "legendOffset": -40,
        "legendPosition": "middle",
        "tickPadding": 5,
        "tickRotation": 0,
        "tickSize": 5,
      }
    }
    axisRight={null}
    axisTop={null}
    colors="set3"
    curve="natural"
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "x": 1614803627951,
              "y": 0,
            },
          ],
          "id": "% Test",
        },
      ]
    }
    dotBorderColor="#fff"
    dotBorderWidth={1}
    dotColor="inherit:darker(0.3)"
    dotSize={8}
    enableArea={true}
    enableDotLabel={true}
    enableStackTooltip={true}
    isInteractive={true}
    margin={
      Object {
        "bottom": 200,
        "left": 75,
        "right": 50,
        "top": 25,
      }
    }
    motionDamping={15}
    motionStiffness={90}
    theme={
      Object {
        "axis": Object {
          "domain": Object {
            "line": Object {
              "stroke": "transparent",
              "strokeWidth": 1,
            },
          },
          "legend": Object {
            "text": Object {
              "fill": "#bbb",
              "fontSize": 12,
            },
          },
          "ticks": Object {
            "line": Object {
              "stroke": "#fff",
              "strokeWidth": 1,
            },
            "text": Object {
              "fill": "#bbb",
              "fontSize": 12,
            },
          },
        },
        "background": "transparent",
        "dots": Object {
          "text": Object {
            "fill": "#bbb",
            "fontSize": 12,
          },
        },
        "grid": Object {
          "line": Object {
            "stroke": "#ddd",
            "strokeWidth": 1,
          },
        },
        "labels": Object {
          "text": Object {
            "fill": "#bbb",
            "fontSize": 12,
          },
        },
        "legends": Object {
          "text": Object {
            "fill": "#bbb",
            "fontSize": 12,
          },
        },
        "markers": Object {
          "fontSize": "12",
          "lineColor": "#fff",
          "lineStrokeWidth": 1,
          "textColor": "#999",
        },
        "tooltip": Object {
          "basic": Object {
            "alignItems": "center",
            "display": "flex",
            "whiteSpace": "pre",
          },
          "container": Object {
            "background": "#222",
            "borderRadius": "2px",
            "boxShadow": "0 1px 2px rgba(255, 255, 255, 0.25)",
            "color": "inherit",
            "fontSize": "inherit",
            "padding": "5px 9px",
          },
          "table": Object {},
          "tableCell": Object {
            "padding": "3px 5px",
          },
        },
      }
    }
    xScale={
      Object {
        "format": "%Q",
        "precision": "minute",
        "type": "time",
      }
    }
    yScale={
      Object {
        "max": 100,
        "min": 0,
        "stacked": false,
        "type": "linear",
      }
    }
  />
</div>
`;

exports[`Paginator Shallow component Matches throughput y-axis snapshot 1`] = `
<div
  className="lineGraph col-12 col-md-6 col-xl-4"
>
  <h5>
    Test
     (Bytes/Minute)
  </h5>
  <ResponsiveLine
    animate={true}
    areaBaselineValue={0}
    areaOpacity={0.5}
    axisBottom={
      Object {
        "format": "%x %H:%M:%S",
        "legend": "Time Stamp",
        "legendOffset": 140,
        "legendPosition": "middle",
        "tickPadding": 5,
        "tickRotation": -90,
        "tickSize": 5,
      }
    }
    axisLeft={
      Object {
        "legend": undefined,
        "legendOffset": -40,
        "legendPosition": "middle",
        "tickPadding": 5,
        "tickRotation": 0,
        "tickSize": 5,
      }
    }
    axisRight={null}
    axisTop={null}
    colors="set3"
    curve="natural"
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "x": 1614803627951,
              "y": 0,
            },
          ],
          "id": "Test",
        },
      ]
    }
    dotBorderColor="#fff"
    dotBorderWidth={1}
    dotColor="inherit:darker(0.3)"
    dotSize={8}
    enableArea={true}
    enableDotLabel={true}
    enableStackTooltip={true}
    isInteractive={true}
    margin={
      Object {
        "bottom": 200,
        "left": 75,
        "right": 50,
        "top": 25,
      }
    }
    motionDamping={15}
    motionStiffness={90}
    theme={
      Object {
        "axis": Object {
          "domain": Object {
            "line": Object {
              "stroke": "transparent",
              "strokeWidth": 1,
            },
          },
          "legend": Object {
            "text": Object {
              "fill": "#bbb",
              "fontSize": 12,
            },
          },
          "ticks": Object {
            "line": Object {
              "stroke": "#fff",
              "strokeWidth": 1,
            },
            "text": Object {
              "fill": "#bbb",
              "fontSize": 12,
            },
          },
        },
        "background": "transparent",
        "dots": Object {
          "text": Object {
            "fill": "#bbb",
            "fontSize": 12,
          },
        },
        "grid": Object {
          "line": Object {
            "stroke": "#ddd",
            "strokeWidth": 1,
          },
        },
        "labels": Object {
          "text": Object {
            "fill": "#bbb",
            "fontSize": 12,
          },
        },
        "legends": Object {
          "text": Object {
            "fill": "#bbb",
            "fontSize": 12,
          },
        },
        "markers": Object {
          "fontSize": "12",
          "lineColor": "#fff",
          "lineStrokeWidth": 1,
          "textColor": "#999",
        },
        "tooltip": Object {
          "basic": Object {
            "alignItems": "center",
            "display": "flex",
            "whiteSpace": "pre",
          },
          "container": Object {
            "background": "#222",
            "borderRadius": "2px",
            "boxShadow": "0 1px 2px rgba(255, 255, 255, 0.25)",
            "color": "inherit",
            "fontSize": "inherit",
            "padding": "5px 9px",
          },
          "table": Object {},
          "tableCell": Object {
            "padding": "3px 5px",
          },
        },
      }
    }
    xScale={
      Object {
        "format": "%Q",
        "precision": "minute",
        "type": "time",
      }
    }
    yScale={
      Object {
        "max": "auto",
        "min": "auto",
        "stacked": false,
        "type": "linear",
      }
    }
  />
</div>
`;
